src/server/replica.cc

```diff
@@ -438,33 +446,66 @@ error_code Replica::InitiateDflySync() {
   // Blocked on until all flows got full sync cut.
   fibers_ext::BlockingCounter sync_block{num_df_flows_};
 
+  // Switch to new error handler that closes flow sockets.
   auto err_handler = [this, sync_block](const auto& ge) mutable {
-    sync_block.Cancel();      // Unblock this function.
-    DefaultErrorHandler(ge);  // Close sockets to unblock flows.
+    // Unblock this function.
+    sync_block.Cancel();
+
+    // Make sure the flows are not in a state transition
+    lock_guard lk{flows_op_mu_};
+
+    // Unblock all sockets.
+    DefaultErrorHandler(ge);
+    for (auto& flow : shard_flows_)
+      flow->CloseSocket();
   };
   RETURN_ON_ERR(cntx_.Switch(std::move(err_handler)));
 
+  // Make sure we're in LOADING state.
+  // TODO: Flush db on retry.
+  CHECK(service_.SwitchState(GlobalState::ACTIVE, GlobalState::LOADING) == GlobalState::LOADING);
+
   // Start full sync flows.
-  auto partition = Partition(num_df_flows_);
-  shard_set->pool()->AwaitFiberOnAll([&](unsigned index, auto*) {
-    for (auto id : partition[index]) {
-      auto ec = shard_flows_[id]->StartFullSyncFlow(sync_block, &cntx_);
-      if (ec)
-        cntx_.Error(ec);
-    }
-  });
+  {
+    auto partition = Partition(num_df_flows_);
+    auto shard_cb = [&](unsigned index, auto*) {
+      for (auto id : partition[index]) {
+        auto ec = shard_flows_[id]->StartFullSyncFlow(sync_block, &cntx_);
+        if (ec)
+          cntx_.Error(ec);
+      }
+    };
+
+    // Lock to prevent the error handler from running instantly
+    // while the flows are in a mixed state.
+    lock_guard lk{flows_op_mu_};
```

Comment:
Is there a deadlock here?
You call AwaitFiberOnAll with shard_cb that can invoke error (line 475).
The Error call will call ReportInternal which will wait for error handler to finish, but you want to prevent the error handler to run here

Reply:
No, because the error handler is launched as a separate fiber. It will suspend and wake up once the initialization is done

Reply:
there is a call to CheckHandlerFb to join the fiber.